### PR TITLE
Add compile flag to disable compilation of mining code

### LINKED ISSUE #1795: Add config flags to disable inclusion of the miner
We want the default zcashd binary to be as usable as possible for causal users to be able to participate in the network. Part of this includes providing usable miners in the binary, so that users can install-and-go. But some users may not want the additional bloat - in particular, mining pool operators. We should add a configuration flag to disable inclusion of the mining code at compile-time, to reduce the size of the binaries for those who don't need it.

### PR DESCRIPTION
**Pull request by str4d**
Closes #1795

### CODE REVIEW

**Comment by str4d** [comment]
This currently works in the disabling direction, but compilation of `libzcashconsensus` fails when the miner is enabled:

```
/bin/bash ../libtool  --tag=CXX   --mode=compile /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/share/../native/bin/ccache g++ -m64 -DHAVE_CONFIG_H -I. -I../src/config  -I/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include -I./obj -DBUILD_BITCOIN_INTERNAL -I/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/share/../include/libsnark -I/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/share/../include/  -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2  -std=c++11 -pipe -O1 -fwrapv -fno-strict-aliasing -Werror -g  -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -fPIE -MT crypto/libzcashconsensus_la-equihash.lo -MD -MP -MF crypto/.deps/libzcashconsensus_la-equihash.Tpo -c -o crypto/libzcashconsensus_la-equihash.lo `test -f 'crypto/equihash.cpp' || echo './'`crypto/equihash.cpp
libtool: compile:  /home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/share/../native/bin/ccache g++ -m64 -DHAVE_CONFIG_H -I. -I../src/config -I/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/include -I./obj -DBUILD_BITCOIN_INTERNAL -I/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/share/../include/libsnark -I/home/str4d/dev/zcash/zcash/depends/x86_64-unknown-linux-gnu/share/../include/ -DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -std=c++11 -pipe -O1 -fwrapv -fno-strict-aliasing -Werror -g -Wformat -Wformat-security -Wstack-protector -fstack-protector-all -MT crypto/libzcashconsensus_la-equihash.lo -MD -MP -MF crypto/.deps/libzcashconsensus_la-equihash.Tpo -c crypto/equihash.cpp  -fPIC -DPIC -o crypto/.libs/libzcashconsensus_la-equihash.o
crypto/equihash.cpp:326:88: error: no ‘bool Equihash<N, K>::BasicSolve(const eh_HashState&, std::function<bool(std::vector<unsigned char>)>, std::function<bool(EhSolverCancelCheck)>)’ member function declared in class ‘Equihash<N, K>’
                                const std::function<bool(EhSolverCancelCheck)> cancelled)
                                                                                        ^
crypto/equihash.cpp:497:92: error: no ‘bool Equihash<N, K>::OptimisedSolve(const eh_HashState&, std::function<bool(std::vector<unsigned char>)>, std::function<bool(EhSolverCancelCheck)>)’ member function declared in class ‘Equihash<N, K>’
                                    const std::function<bool(EhSolverCancelCheck)> cancelled)
                                                                                            ^
crypto/equihash.cpp:770:98: error: no member function ‘BasicSolve’ declared in ‘Equihash<96u, 3u>’
                                          const std::function<bool(EhSolverCancelCheck)> cancelled);
                                                                                                  ^
crypto/equihash.cpp:773:102: error: no member function ‘OptimisedSolve’ declared in ‘Equihash<96u, 3u>’
                                              const std::function<bool(EhSolverCancelCheck)> cancelled);
                                                                                                      ^
crypto/equihash.cpp:782:99: error: no member function ‘BasicSolve’ declared in ‘Equihash<200u, 9u>’
                                           const std::function<bool(EhSolverCancelCheck)> cancelled);
                                                                                                   ^
crypto/equihash.cpp:785:103: error: no member function ‘OptimisedSolve’ declared in ‘Equihash<200u, 9u>’
                                               const std::function<bool(EhSolverCancelCheck)> cancelled);
                                                                                                       ^
crypto/equihash.cpp:794:98: error: no member function ‘BasicSolve’ declared in ‘Equihash<96u, 5u>’
                                          const std::function<bool(EhSolverCancelCheck)> cancelled);
                                                                                                  ^
crypto/equihash.cpp:797:102: error: no member function ‘OptimisedSolve’ declared in ‘Equihash<96u, 5u>’
                                              const std::function<bool(EhSolverCancelCheck)> cancelled);
                                                                                                      ^
crypto/equihash.cpp:806:98: error: no member function ‘BasicSolve’ declared in ‘Equihash<48u, 5u>’
                                          const std::function<bool(EhSolverCancelCheck)> cancelled);
                                                                                                  ^
crypto/equihash.cpp:809:102: error: no member function ‘OptimisedSolve’ declared in ‘Equihash<48u, 5u>’
                                              const std::function<bool(EhSolverCancelCheck)> cancelled);
                                                                                                      ^
Makefile:6093: recipe for target 'crypto/libzcashconsensus_la-equihash.lo' failed
make[2]: *** [crypto/libzcashconsensus_la-equihash.lo] Error 1
```

It looks like the `ENABLE_MINER` flag is being passed to `equihash.cpp`, but not to the included `equihash.h`...

File: src/crypto/equihash.h [hunk 1]
```diff
@@ -89,17 +101,46 @@ class FullStepRow : public StepRow<WIDTH>
...
+        EhIndexToArray(i, hash+hLen);
+    }
     ~FullStepRow() { }
 
     FullStepRow(const FullStepRow<WIDTH>& a) : StepRow<WIDTH> {a} { }
+
     template<size_t W>
-    FullStepRow(const FullStepRow<W>& a, const FullStepRow<W>& b, size_t len, size_t lenIndices, int trim);
-    FullStepRow& operator=(const FullStepRow<WIDTH>& a);
+    FullStepRow(const FullStepRow<W>& a, const FullStepRow<W>& b, size_t len, size_t lenIndices, int trim) :
```
Comment (ebfull) [inline src/crypto/equihash.h]: What is all this stuff being changed that doesn't have to do with conditional compilation of the mining code?

**Comment by ebfull** [comment]
There's too much happening in this PR that isn't explained... not to say any of it is wrong, but that I really don't know why about 80% of the code changes are there.

**Comment by str4d** [comment]
@ebfull Here's what is going on:

- The first commit adds the compile flag and does the conditional compilation of the mining code. However, it doesn't compile because the conditional exclusion of the default solver causes various template errors with mining code *enabled* that I was unable to solve.
- The second commit extracts the default solver into a standalone subclass, and then conditionally compiles *that*. This I was able to get working.

**Comment by bitcartel** [comment]
There are some PRs dependent upon this one, with some blocked for various reasons.  This PR on its own is not a blocker for 1.0.5 since users by default launch zcashd with mining disabled.  If the other PRs are not likely to be resolved in time for todays tagging, building and packaging, this PR could be delayed for 1.0.6 giving is time to figure out why the first commit can lead to compile errors.  @str4d Is the second commit specific to this PR or would we have done it anyway when adding more solvers?

**Comment by str4d** [comment]
@bitcartel AIUI the eventual goal is to have a unified interface to any underlying solvers. That *could* be done without the second commit though, and if the compile-time errors could be fixed then I'd be fine with removing the second commit.

**Comment by str4d** [comment]
@ebfull @bitcartel I figured out the cause of the compile error, and have removed the code-moving commit, so this PR is much simpler now.

**Comment by bitcartel** [comment]
@zkbot try

**Comment by bitcartel** [comment]
@str4d compile error
```
rpcmining.cpp: In function ‘json_spirit::Value generate(const Array&, bool)’:
rpcmining.cpp:271:1: error: a function-definition is not allowed here before ‘{’ token
 {
 ^
rpcmining.cpp:905:1: error: expected ‘}’ at end of input
 }
 ^
Makefile:4250: recipe for target 'libbitcoin_server_a-rpcmining.o' failed
make[2]: *** [libbitcoin_server_a-rpcmining.o] Error 1
```

**Comment by bitcartel** [comment]
@str4d @ebfull I think this is good to go (see above).

**Comment by daira** [comment]
This is a subset of https://github.com/zcash/zcash/pull/1965 , which I've utACKed.

**Comment by bitcartel** [comment]
@zbkot r+

**Comment by bitcartel** [comment]
@zkbot r+

**Pull request merged**